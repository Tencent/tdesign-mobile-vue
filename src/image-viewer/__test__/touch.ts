/**
 * https://github.com/vant-ui/vant/blob/dev/packages/vant/test/event.ts
 */

import { ComponentPublicInstance, nextTick } from 'vue';
import { VueWrapper, DOMWrapper } from '@vue/test-utils';

function getTouch(el: Element | Window, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  };
}

// trigger pointer/touch event
export function trigger(
  wrapper: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | Element | Window,
  eventName: string,
  x = 0,
  y = 0,
  options: any = {},
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper;
  const touchList = options.touchList || [getTouch(el, x, y)];

  if (options.x || options.y) {
    touchList.push(getTouch(el, options.x, options.y));
  }

  const event = document.createEvent('CustomEvent');
  event.initCustomEvent(eventName, true, true, {});

  Object.assign(event, {
    clientX: x,
    clientY: y,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
  });

  el.dispatchEvent(event);

  return nextTick();
}

function triggerTwoFingerTouchmove(el: Element | DOMWrapper<Element>, x: number, y: number) {
  trigger(el, 'touchmove', -x, -y, { x, y });
}

// simulate zoom
export function triggerZoom(el: Element | DOMWrapper<Element>, x: number, y: number, direction: 'in' | 'out' = 'in') {
  trigger(el, 'touchstart', 0, 0, { x, y });

  if (direction === 'in') {
    triggerTwoFingerTouchmove(el, x / 4, y / 4);
    triggerTwoFingerTouchmove(el, x / 3, y / 3);
    triggerTwoFingerTouchmove(el, x / 2, y / 2);
    triggerTwoFingerTouchmove(el, x, y);
  } else if (direction === 'out') {
    triggerTwoFingerTouchmove(el, x, y);
    triggerTwoFingerTouchmove(el, x / 2, y / 2);
    triggerTwoFingerTouchmove(el, x / 3, y / 3);
    triggerTwoFingerTouchmove(el, x / 4, y / 4);
  }

  trigger(el, 'touchend', 0, 0, { touchList: [] });
}
