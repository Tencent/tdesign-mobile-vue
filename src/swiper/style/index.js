import '../../_common/style/mobile/components/swiper/v2/_index.less';
import '../../_common/style/mobile/components/swiper-nav/v2/_index.less';
