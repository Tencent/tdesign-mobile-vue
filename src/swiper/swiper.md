:: BASE_DOC ::

## API
### Swiper Props

名称 | 类型 | 默认值 | 说明 | 必传
-- | -- | -- | -- | --
animation | String | slide | 轮播切换动画效果类型。可选项：slide | N
autoplay | Boolean | true | 是否自动播放 | N
current | Number | 0 | 当前轮播在哪一项（下标）。支持语法糖 `v-model` 或 `v-model:current` | N
direction | String | horizontal | 轮播滑动方向，包括横向滑动和纵向滑动两个方向。可选项：horizontal/vertical | N
duration | Number | 300 | 滑动动画时长 | N
height | String / Number | 192 | 已废弃。轮播的高度；默认单位 `px` | N
interval | Number | 5000 | 轮播间隔时间 | N
loop | Boolean | true | 是否循环播放 | N
navigation | Object / Slot / Function | - | 导航器全部配置。TS 类型：`SwiperNavigation \| TNode`。[通用类型定义](https://github.com/Tencent/tdesign-mobile-vue/blob/develop/src/common.ts) | N
nextAffset | String / Number | 0 | 后边距偏移量，可用于露出后一项的一小部分。默认单位 `px` | N
previousAffset | String / Number | 0 | 前边距偏移量，可用于露出前一项的一小部分。默认单位 `px` | N
onChange | Function |  | TS 类型：`(current: number, context: { source: SwiperChangeSource }) => void`<br/>轮播切换时触发。[详细类型定义](https://github.com/Tencent/tdesign-mobile-vue/tree/develop/src/swiper/type.ts)。<br/>`type SwiperChangeSource = 'autoplay' \| 'touch' \| ''`<br/> | N
onClick | Function |  | TS 类型：`(index: number) => void`<br/>点击轮播项时触发 | N

### Swiper Events

名称 | 参数 | 描述
-- | -- | --
change | `(current: number, context: { source: SwiperChangeSource })` | 轮播切换时触发。[详细类型定义](https://github.com/Tencent/tdesign-mobile-vue/tree/develop/src/swiper/type.ts)。<br/>`type SwiperChangeSource = 'autoplay' \| 'touch' \| ''`<br/>
click | `(index: number)` | 点击轮播项时触发

### SwiperNavigation

名称 | 类型 | 默认值 | 说明 | 必传
-- | -- | -- | -- | --
current | Number | 0 | 当前轮播在哪一项（下标） | N
minShowNum | Number | - | 小于这个数字不会显示导航器 | N
paginationPosition | String | bottom | 页码信息展示位置。可选项：top-left/top/top-right/bottom-left/bottom/bottom-right | N
placement | String | inside | 导航器位置，位于主体的内侧或是外侧。可选项：inside/outside | N
showControls | Boolean | false | 是否显示两侧的控制按钮 | N
total | Number | 0 | 总共的项数 | N
type | String | dots | 导航器类型，点状(dots)、点条状(dots-bar)、分式(fraction)等。TS 类型：`SwiperNavigationType` `type SwiperNavigationType = 'dots' \| 'dots-bar' \| 'fraction'`。[详细类型定义](https://github.com/Tencent/tdesign-mobile-vue/tree/develop/src/swiper/type.ts) | N
