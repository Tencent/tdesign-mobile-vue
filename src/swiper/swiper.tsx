import { onMounted, computed, ref, provide, watch, onUnmounted, toRefs, defineComponent } from 'vue';
import isObject from 'lodash/isObject';
import isNumber from 'lodash/isNumber';
import { useSwipe } from '../swipe-cell/useSwipe';

import config from '../config';
import props from './props';
import { SwiperChangeSource, SwiperNavigation } from './type';
import { useVModel } from '../shared';
import { preventDefault } from '../shared/dom';
import { useTNodeJSX } from '../hooks/tnode';
import { usePrefixClass } from '../hooks/useClass';

const { prefix } = config;
const name = `${prefix}-swiper`;
const navName = `${prefix}-swiper-nav`;
export default defineComponent({
  name,
  props,
  emits: ['change', 'update:current', 'update:modelValue'],
  setup(props, context) {
    const swiperClass = usePrefixClass('swiper');
    const readerTNodeJSX = useTNodeJSX();
    const setOffset = (offset: number, direction = 'X'): void => {
      translateContainer.value = `translate${direction}(${offset}px)`;
    };

    const root = ref();
    const items = ref<any>([]);

    const { current: value, modelValue } = toRefs(props);

    const [current, setCurrent] = useVModel(value, modelValue, props.defaultCurrent);
    const swiperContainer = ref<HTMLElement | null>(null);

    const animating = ref(false);
    const disabled = ref(false);
    const isSwiperDisabled = computed(() => props.disabled === true);
    const translateContainer = ref('');

    const isVertical = computed(() => props.direction === 'vertical');
    const containerHeight = ref('auto');

    const navigation = computed((): SwiperNavigation => props.navigation);

    const isBottomPagination = computed(() => {
      let isShowSwiperNav = false;
      if (typeof props.navigation === 'object') {
        isShowSwiperNav =
          (!navigation.value?.paginationPosition || navigation.value?.paginationPosition === 'bottom') &&
          (navigation.value?.type === 'dots' || navigation.value?.type === 'dots-bar') &&
          enableNavigation?.value;
      }
      return isShowSwiperNav;
    });

    const rootClass = computed(() => {
      return [
        `${swiperClass.value}`,
        `${swiperClass.value}--${props.type}`,
        `${
          isBottomPagination.value && navigation.value?.placement
            ? `${swiperClass.value}--${navigation.value?.placement}`
            : ''
        }`,
      ];
    });

    const enableNavigation = computed(() => {
      if (typeof props.navigation === 'object') {
        return navigation.value?.minShowNum ? items.value.length >= navigation.value?.minShowNum : true;
      }
      return false;
    });

    let autoplayTimer: any = null;

    const onItemClick = () => {
      props.onClick?.(current.value ?? 0);
    };

    const move = (step: number, source: SwiperChangeSource, isReset = false) => {
      animating.value = true;
      processIndex(isReset ? step : (current.value as number) + step, source);

      const moveDirection = !isVertical.value ? 'X' : 'Y';
      const distance = root.value?.[isVertical.value ? 'offsetHeight' : 'offsetWidth'] ?? 0;

      translateContainer.value = `translate${moveDirection}(${isReset ? 0 : -1 * distance * step}px)`;
    };

    const handleAnimationEnd = () => {
      disabled.value = false;
      animating.value = false;
      translateContainer.value = 'translateX(0)';

      updateItemPosition();
    };

    const stopAutoplay = () => {
      if (!autoplayTimer) return;
      clearInterval(autoplayTimer as number);
      autoplayTimer = null;
    };

    const startAutoplay = () => {
      if (typeof props.current === 'number') return false;
      if (!props?.autoplay || autoplayTimer !== null) return false; // stop repeat autoplay
      autoplayTimer = setInterval(() => {
        goNext('autoplay');
      }, props?.interval);
    };

    const goPrev = (source: SwiperChangeSource) => {
      disabled.value = true;
      move(-1, source);
    };
    const goNext = (source: SwiperChangeSource) => {
      disabled.value = true;
      move(1, source);
    };

    const processIndex = (index: number, source: SwiperChangeSource) => {
      const max = items.value.length;
      let val = index;

      if (index < 0) {
        val = props.loop ? max - 1 : 0;
      }
      if (index >= max) {
        val = props.loop ? 0 : max - 1;
      }
      setCurrent(val);
      context.emit('change', val, { source });
    };

    const { lengthX, lengthY } = useSwipe(swiperContainer, {
      onSwipeStart() {
        if (disabled.value || isSwiperDisabled.value || !items.value.length) return;
        stopAutoplay();
      },
      onSwipe(e: TouchEvent) {
        if (disabled.value || isSwiperDisabled.value || !items.value.length) return;
        onTouchMove(e);
      },
      onSwipeEnd() {
        if (disabled.value || isSwiperDisabled.value || !items.value.length) return;
        onTouchEnd();
      },
    });

    const onTouchMove = (event: TouchEvent) => {
      preventDefault(event, false);
      const distanceX = lengthX.value;
      const distanceY = lengthY.value;

      animating.value = false;

      if (!isVertical.value) {
        setOffset(-distanceX);
      } else {
        setOffset(-distanceY, 'Y');
      }
    };

    const onTouchEnd = () => {
      const distanceX = lengthX.value;
      const distanceY = lengthY.value;

      if ((!isVertical.value && distanceX < -100) || (isVertical.value && distanceY < -100)) {
        move(-1, 'touch');
      } else if ((!isVertical.value && distanceX > 100) || (isVertical.value && distanceY > 100)) {
        move(1, 'touch');
      } else {
        move(current.value as number, 'touch', true);
      }
      startAutoplay();
    };

    const addChild = (item: any) => {
      items.value.push(item);
    };

    const removeChild = (uid: number) => {
      const index = items.value.findIndex((item: any) => item.uid === uid);
      items.value.splice(index, 1);

      if (current.value + 1 > items.value.length) {
        goNext('autoplay');
      }
    };

    const updateItemPosition = () => {
      items.value.forEach((item: any, index: number) => {
        item.calcTranslateStyle(index, current.value);
      });
    };

    const setContainerHeight = (height: number | string) =>
      (containerHeight.value = isNumber(height) ? `${height}px` : height);

    const updateContainerHeight = () => {
      const target = items.value[current.value ?? 0];
      const rect = target?.proxy?.$el.getBoundingClientRect();

      if (props.height) {
        setContainerHeight(props.height);
      } else if (rect) {
        setContainerHeight(rect.height);
      }
      updateItemPosition();
    };

    watch(current, updateContainerHeight);

    provide('parent', {
      loop: props.loop,
      root,
      items,
      isVertical,
      addChild,
      removeChild,
      setContainerHeight,
    });

    onMounted(() => {
      startAutoplay();
      updateItemPosition();
      updateContainerHeight();
    });

    onUnmounted(() => {
      stopAutoplay();
    });
    return () => {
      const swiperNav = () => {
        if (navigation.value && enableNavigation.value) {
          const controlsNav = () => {
            if (!isVertical.value && !!navigation.value?.showControls) {
              return (
                <span class={`${navName}__btn`}>
                  <span class={`${navName}__btn--prev`} onClick={() => goPrev('nav')} />
                  <span class={`${navName}__btn--next`} onClick={() => goNext('nav')} />
                </span>
              );
            }
          };
          const typeNav = () => {
            if ('type' in navigation.value) {
              // dots
              const dots = () => {
                if (['dots', 'dots-bar'].includes(navigation.value?.type || '')) {
                  return (
                    <>
                      {items.value.map((_: any, index: number) => (
                        <span
                          key={`page${index}`}
                          class={[
                            `${navName}__${navigation.value?.type}-item`,
                            index === current.value ? `${navName}__${navigation.value?.type}-item--active` : '',
                            `${navName}__${navigation.value?.type}-item--${props.direction}`,
                          ]}
                        />
                      ))}
                    </>
                  );
                }
              };
              // fraction
              const fraction = () => {
                if (navigation.value?.type === 'fraction') {
                  return <span>{`${(current.value ?? 0) + 1}/${items.value.length}`}</span>;
                }
              };
              return (
                <span
                  class={[
                    `${navName}--${props.direction}`,
                    `${navName}__${navigation.value?.type || ''}`,
                    `${navName}--${navigation.value?.paginationPosition || 'bottom'}`,
                    `${
                      isBottomPagination.value && navigation.value?.placement
                        ? `${navName}--${navigation.value?.placement}`
                        : ''
                    }`,
                  ]}
                >
                  {dots()}
                  {fraction()}
                </span>
              );
            }
          };

          return (
            <>
              {controlsNav()}
              {typeNav()}
            </>
          );
        }
        return isObject(props.navigation) ? '' : readerTNodeJSX('navigation');
      };
      return (
        <div ref={root} class={rootClass.value}>
          <div
            ref={swiperContainer}
            class={`${swiperClass.value}__container`}
            style={{
              flexDirection: !isVertical.value ? 'row' : 'column',
              transition: animating.value ? `transform ${props.duration}ms` : 'none',
              transform: translateContainer.value,
              height: containerHeight.value,
            }}
            onTransitionend={(event: TransitionEvent) => {
              if (event.target === event.currentTarget) {
                handleAnimationEnd();
              }
            }}
            onClick={onItemClick}
          >
            {readerTNodeJSX('default')}
          </div>
          {swiperNav()}
        </div>
      );
    };
  },
});
