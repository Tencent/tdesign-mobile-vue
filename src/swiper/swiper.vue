<template>
  <div :style="{ height: `${height}px`, overflow: 'hidden' }" :class="`${name}`">
    <div
      ref="swiperContainer"
      :class="`${name}__container`"
      :style="{
        height: `${height}px`,
        flexDirection: direction === 'horizontal' ? 'row' : 'column',
      }"
      @transitionend="handleAnimationEnd"
    >
      <slot></slot>
    </div>
    <template v-if="navigation && state.showNavigation">
      <span v-if="direction === 'horizontal' && 'showSlideBtn' in navigation && navigation.showSlideBtn">
        <span :class="`${name}__btn btn-prev`" @click="prev(1)">
          <chevron-left-icon size="20px" />
        </span>
        <span :class="`${name}__btn btn-next`" @click="next(1)">
          <chevron-right-icon size="20px" />
        </span>
      </span>
      <span
        v-if="'type' in navigation"
        :class="`${name}__pagination ${name}__pagination-${
          navigation.type || ''
        } ${name}__pagination-${paginationPosition}`"
      >
        <template v-if="['dots', 'dots-bar'].includes(navigation.type || '')">
          <span
            v-for="(item, index) in state.children.length"
            :key="'page' + index"
            :class="{ [`${name}-dot`]: true, [`${name}-dot--active`]: index === state.activeIndex }"
          ></span>
        </template>
        <span v-if="navigation.type && navigation.type === 'fraction'">
          {{ showPageNum + '/' + state.children.length }}
        </span>
      </span>
    </template>
    <template v-else-if="computedNavigation !== undefined">
      <t-node :content="computedNavigation" :style="{}"></t-node>
    </template>
  </div>
</template>
<script lang="ts">
import {
  defineComponent,
  reactive,
  getCurrentInstance,
  onMounted,
  computed,
  watch,
  ref,
  SetupContext,
  nextTick,
  provide,
  ComponentInternalInstance,
  ComponentPublicInstance,
} from 'vue';
import { ChevronLeftIcon, ChevronRightIcon } from 'tdesign-icons-vue-next';
import { useSwipe } from '@vueuse/core';
import SwiperProps from './props';
import config from '../config';
import { renderTNode, useDefault, TNode } from '../shared';
import { TdSwiperProps } from './type';

const { prefix } = config;
const name = `${prefix}-swiper`;
const setOffset = (element: HTMLDivElement, offset: number, direction = 'X'): void => {
  // eslint-disable-next-line no-param-reassign
  element.style.transform = `translate${direction}(${offset}px)`;
};
export default defineComponent({
  name,
  components: { ChevronLeftIcon, ChevronRightIcon, TNode },
  props: SwiperProps,
  emits: ['change', 'update:current', 'update:modelValue'],
  setup(props, context: SetupContext) {
    // const emitEvent = useEmitEvent(props, context.emit);
    const [swiperValue, setSwiperValue] = useDefault<Number, TdSwiperProps>(props, context.emit, 'current', 'change');
    const self = getCurrentInstance();
    const swiperContainer = ref<HTMLElement | null>(null);
    const computedNavigation = computed(() => renderTNode(self, 'navigation'));
    // const { height = 180, current = null } = props;
    const height = props.height || 180;
    const state: {
      showNavigation: boolean;
      activeIndex: number;
      itemLength: number;
      itemWidth: number;
      isControl: boolean;
      btnDisabled: boolean;
      children: ComponentPublicInstance[];
    } = reactive({
      showNavigation: true,
      activeIndex: 0,
      itemLength: 0,
      itemWidth: 0,
      isControl: false,
      btnDisabled: false,
      children: [] as ComponentPublicInstance[],
    });
    const paginationList = computed(() => new Array(state.itemLength).fill(1));
    const showPageNum = computed(() => {
      const { activeIndex, itemLength } = state;
      if (activeIndex > itemLength - 1) return itemLength;
      if (activeIndex < 0) return 1;
      return activeIndex + 1;
    });
    const childCount = computed(() => state.children.length);
    const getContainer = (): HTMLDivElement => self?.proxy?.$el.querySelector(`.${name}__container`);
    // const getContainer = (): HTMLDivElement => swiperContainer.value as any;
    const initSwiper = () => {
      const _swiperContainer = getContainer();
      _swiperContainer.querySelectorAll('.copy-item').forEach((ele) => {
        _swiperContainer.removeChild(ele);
      });
      const items = _swiperContainer.querySelectorAll(`.${name}-item`);
      state.itemLength = _swiperContainer.children?.length || 0;
      const itemWidth = _swiperContainer.querySelector(`.${name}-item`)?.getBoundingClientRect().width || 0;
      state.itemWidth = itemWidth;
      if (items.length <= 0) return false;
      if (
        computedNavigation.value &&
        'minShowNum' in computedNavigation.value &&
        items.length < computedNavigation.value.minShowNum
      ) {
        state.showNavigation = false;
      }
      if (props?.loop) {
        const first = items[0].cloneNode(true) as HTMLDivElement;
        first.classList.add('copy-item');
        const last = items[items.length - 1].cloneNode(true) as HTMLDivElement;
        last.classList.add('copy-item');
        _swiperContainer.appendChild(first);
        _swiperContainer.insertBefore(last, items[0]);
      }
      move(0);
      startAutoplay();
      if (typeof props.current === 'number') {
        state.isControl = true;
        next(props.current);
      }
    };
    onMounted(() => {
      nextTick(() => {
        console.info('swiper mounted');
        initSwiper();
      });
    });
    watch(
      () => state.children.length,
      () => {
        nextTick(() => {
          console.info('swiper mounted');
          initSwiper();
        });
      },
    );
    // eslint-disable-next-line no-undef
    let autoplayTimer: number | NodeJS.Timeout | null = null;
    let actionIsTrust = true;
    /**
     * move item
     */
    const move = (targetIndex: number, isTrust = true) => {
      const _swiperContainer = getContainer();
      const moveDirection = props?.direction === 'horizontal' ? 'X' : 'Y';
      const moveLength: number = props?.direction === 'vertical' ? height : state.itemWidth;
      actionIsTrust = isTrust;
      _swiperContainer.dataset.isTrust = `${isTrust}`;
      // do not translate one item if not loop
      const toIndex = props?.loop ? targetIndex + 1 : targetIndex;
      _swiperContainer.style.transform = `translate${moveDirection}(-${moveLength * toIndex}px)`;
    };
    const addAnimation = () => {
      const _swiperContainer = getContainer();
      _swiperContainer.style.transition = `transform ${props?.duration}ms`;
    };
    const removeAnimation = () => {
      const _swiperContainer = getContainer();
      _swiperContainer.style.transition = 'none';
    };
    const handleAnimationEnd = () => {
      state.btnDisabled = false;
      removeAnimation();
      if (state.activeIndex >= state.itemLength) {
        state.activeIndex = 0;
        move(0);
      }
      if (state.activeIndex <= -1) {
        state.activeIndex = state.itemLength - 1;
        move(state.itemLength - 1);
      }
      setTimeout(() => {
        actionIsTrust && emitCurrentChange(state.activeIndex);
      }, 0);
    };
    const stopAutoplay = () => {
      if (!autoplayTimer) return;
      clearInterval(autoplayTimer as number);
      autoplayTimer = null;
    };
    const startAutoplay = () => {
      if (typeof props.current === 'number') return false;
      if (!props?.autoplay || autoplayTimer !== null) return false; // stop repeat autoplay
      autoplayTimer = setInterval(() => {
        state.activeIndex += 1;
        if (!props?.loop && state.activeIndex >= state.children.length - 1) {
          state.activeIndex = 0;
        }
        if (!props?.loop && state.activeIndex <= 0) {
          state.activeIndex = state.children.length - 1;
        }
        addAnimation();
        move(state.activeIndex);
      }, props?.interval);
    };
    const emitCurrentChange = (index: number) => {
      // if (!state.isControl) return false;
      let resultIndex = index;
      if (index >= state.itemLength) resultIndex = 0;
      if (index < 0) resultIndex = state.itemLength - 1;
      // emitEvent('change', resultIndex);
      setSwiperValue(resultIndex);
    };
    const prev = (step = 1) => {
      const cannotMovePrev = !props?.loop && state.activeIndex === 0;
      if (state.btnDisabled || cannotMovePrev) {
        move(state.activeIndex);
        return false;
      }
      stopAutoplay();
      state.activeIndex -= step;
      addAnimation();
      move(state.activeIndex);
      startAutoplay();
      state.btnDisabled = true;
    };
    const next = (step = 1) => {
      const cannotMoveLast = !props?.loop && state.activeIndex === state.itemLength - 1;
      if (state.btnDisabled || cannotMoveLast) {
        move(state.activeIndex);
        return false;
      }
      stopAutoplay();
      state.activeIndex += step;
      addAnimation();
      move(state.activeIndex);
      startAutoplay();
      state.btnDisabled = true;
    };
    const { lengthX, lengthY } = useSwipe(swiperContainer, {
      passive: false,
      onSwipeStart(e: TouchEvent) {
        if (state.btnDisabled) return false;
        stopAutoplay();
      },
      onSwipe(e: TouchEvent) {
        if (state.btnDisabled) return false;
        onTouchMove(e);
      },
      onSwipeEnd() {
        onTouchEnd();
      },
    });
    const onTouchMove = (event: TouchEvent) => {
      event.preventDefault();
      const { activeIndex, itemWidth } = state;
      const distanceX = lengthX.value;
      const distanceY = lengthY.value;
      const _container = getContainer();
      removeAnimation();
      const toIndex = props?.loop ? activeIndex + 1 : activeIndex;
      if (props?.direction === 'horizontal') {
        setOffset(_container, -(toIndex * itemWidth + distanceX));
      } else {
        const { height = 180 } = props;
        setOffset(_container, -(toIndex * height + distanceY), 'Y');
      }
    };
    const onTouchEnd = () => {
      const distanceX = lengthX.value;
      const distanceY = lengthY.value;
      addAnimation();
      if (
        (props?.direction === 'horizontal' && distanceX < -100) ||
        (props?.direction === 'vertical' && distanceY < -100)
      ) {
        prev(1);
      } else if (
        (props?.direction === 'horizontal' && distanceX > 100) ||
        (props?.direction === 'vertical' && distanceY > 100)
      ) {
        next(1);
      } else {
        move(state.activeIndex);
      }
      startAutoplay();
    };
    const relation = (child: ComponentInternalInstance) => {
      if (child.proxy) {
        state.children.push(child.proxy);
      }
    };
    provide('parent', {
      props,
      relation,
    });
    watch(
      () => props.current,
      (newPage, oldPage) => {
        if (state.isControl) {
          state.activeIndex = newPage || 0;
          addAnimation();
          move(state.activeIndex, false);
        }
      },
    );
    return {
      swiperContainer,
      name,
      computedNavigation,
      onTouchMove,
      onTouchEnd,
      handleAnimationEnd,
      state,
      paginationList,
      showPageNum,
      prev,
      next,
    };
  },
});
</script>
