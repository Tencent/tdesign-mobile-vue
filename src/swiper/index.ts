import _Swiper from './swiper';
import _SwiperItem from './swiper-item';
import { withInstall, WithInstallType } from '../shared';
import { TdSwiperProps } from './type';

import './style';

export * from './type';
export type SwiperProps = TdSwiperProps;

export const Swiper: WithInstallType<typeof _Swiper> = withInstall(_Swiper);
export const SwiperItem: WithInstallType<typeof _SwiperItem> = withInstall(_SwiperItem);
