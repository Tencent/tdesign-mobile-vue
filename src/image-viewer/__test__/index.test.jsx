import { nextTick, ref } from 'vue';
import { mount } from '@vue/test-utils';
import { AddCircleIcon } from 'tdesign-icons-vue-next';
import { describe, it, vi } from 'vitest';
import ImageViewer from '../image-viewer';
import { Swiper } from '../../swiper';
import { trigger, triggerZoom } from './touch';

const images = ref([
  'https://imgcache.qq.com/open_proj/proj_qcloud_v2/rocket_images/1606728019829_yw760ok1jmpbep14i.png',
  'https://imgcache.qq.com/open_proj/proj_qcloud_v2/rocket_images/1606728019829_yw760ok1jmpbep14i.png',
  'https://imgcache.qq.com/open_proj/proj_qcloud_v2/rocket_images/1606728019829_yw760ok1jmpbep14i.png',
]);

describe('ImageViewer', () => {
  describe('props', () => {
    it(':closeBtn', async () => {
      const wrapper = mount(<ImageViewer v-model:images={images.value} visible={true} closeBtn={false} />);

      expect(wrapper.find('.t-image-viewer__nav-close').exists()).toBe(false);
    });

    it(':images', async () => {
      const emptyImages = mount(<ImageViewer visible={true} />);
      expect(emptyImages.find('.t-swiper-item').exists()).toBe(false);

      const wrapper = mount(<ImageViewer v-model:images={images.value} visible={true} />);

      await nextTick();
      expect(wrapper.findAll('.t-swiper-item').length).toBe(images.value.length);

      images.value.pop();
      await nextTick();
      expect(wrapper.findAll('.t-swiper-item').length).toBe(images.value.length);
    });

    // it(':maxZoom', async () => {
    //   const wrapper = mount(<ImageViewer v-model:images={images.value} visible={true} maxZoom={2} />);

    //   const target = wrapper.find('.t-swiper-item');
    //   triggerZoom(target, 200, 200, 'in');
    //   expect(wrapper.vm.imageStyle).toMatchObject({ transform: `scale(${2}, ${2})` });

    //   triggerZoom(target, 200, 200, 'out');
    //   expect(wrapper.vm.imageStyle).not.toHaveProperty('transform');

    //   // 测试 toggleScale
    //   trigger(target, 'touchstart', 50, 50);
    //   trigger(target, 'touchend', 50, 50);

    //   await new Promise((resolve, reject) => {
    //     setTimeout(() => {
    //       resolve();
    //     }, 100); // 小于 TAP_TIME
    //   });
    //   trigger(target, 'touchstart', 50, 50);
    //   trigger(target, 'touchend', 50, 50);
    // });

    it(':visible', async () => {
      const visible = ref(false);
      const wrapper = mount(<ImageViewer v-model:images={images.value} v-model:visible={visible.value} />);

      expect(wrapper.findComponent(Swiper).exists()).toBe(false);
      visible.value = true;
    });
  });

  describe('slots', () => {
    it(':closeBtn', () => {
      const wrapper = mount(
        <ImageViewer v-model:images={images.value} visible={true} closeBtn={() => <AddCircleIcon />} />,
      );

      expect(wrapper.findComponent(AddCircleIcon).exists()).toBe(true);
    });
  });

  describe('event', () => {
    it(':close', async () => {
      const wrapper = mount(<ImageViewer v-model:images={images.value} visible={true} />);

      await wrapper.find('.t-image-viewer__nav-close').trigger('click');
      expect(wrapper.emitted()).toHaveProperty('close');
    });

    it(': delete', async () => {
      const wrapper = mount(<ImageViewer v-model:images={images.value} visible={true} delete-btn={true} />);
      
      await wrapper.find('.t-image-viewer__nav-delete').trigger('click');
      expect(wrapper.emitted()).toHaveProperty('delete');
    });
  });
});
