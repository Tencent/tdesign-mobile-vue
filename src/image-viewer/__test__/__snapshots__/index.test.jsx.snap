// Vitest Snapshot v1

exports[`ImageViewer > function > : onIndexChange 1`] = `
<transition-stub
  class="t-image-viewer"
>
  <div
    class="t-overlay t-overlay--active"
    style="z-index: 1000; transition-duration: 300ms;"
  >
    
    
    <div
      class="t-image-viewer__close-icon"
    >
      <svg
        class="t-icon t-icon-close-circle-filled"
        fill="none"
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
      >
        <path
          d="M15 8A7 7 0 101 8a7 7 0 0014 0zM5.67 4.95L8 7.29l2.33-2.34.7.7L8.7 8l2.34 2.35-.71.7L8 8.71l-2.33 2.34-.7-.7L7.3 8 4.96 5.65l.71-.7z"
          fill="currentColor"
          fill-opacity="0.9"
        />
      </svg>
    </div>
    <div
      class="t-swiper t-image-viewer__swipe"
      style="overflow: hidden;"
    >
      <div
        class="t-swiper__container"
        data-is-trust="true"
        style="flex-direction: row; transform: translateX(-0px); transition: transform 300ms;"
      >
        
        
        <div
          class="t-swiper-item t-image-viewer__swipe-item copy-item"
        >
          
          <img
            class="t-image-viewer__image"
            src="https://imgcache.qq.com/open_proj/proj_qcloud_v2/rocket_images/1606728019829_yw760ok1jmpbep14i.png"
            style="transition-duration: .3s;"
          />
          
        </div>
        <div
          class="t-swiper-item t-image-viewer__swipe-item"
        >
          
          <img
            class="t-image-viewer__image"
            src="https://imgcache.qq.com/open_proj/proj_qcloud_v2/rocket_images/1606728019829_yw760ok1jmpbep14i.png"
            style="transition-duration: .3s;"
          />
          
        </div>
        <div
          class="t-swiper-item t-image-viewer__swipe-item"
        >
          
          <img
            class="t-image-viewer__image"
            src="https://imgcache.qq.com/open_proj/proj_qcloud_v2/rocket_images/1606728019829_yw760ok1jmpbep14i.png"
            style="transition-duration: .3s;"
          />
          
        </div>
        
        
        <div
          class="t-swiper-item t-image-viewer__swipe-item copy-item"
        >
          
          <img
            class="t-image-viewer__image"
            src="https://imgcache.qq.com/open_proj/proj_qcloud_v2/rocket_images/1606728019829_yw760ok1jmpbep14i.png"
            style="transition-duration: .3s;"
          />
          
        </div>
      </div>
      
      <!--v-if-->
      <span
        class="t-swiper__pagination t-swiper__pagination-dots t-swiper__pagination-bottom"
      >
        
        <span
          class="t-swiper-dot"
        />
        <span
          class="t-swiper-dot"
        />
        
        <!--v-if-->
      </span>
      
    </div>
    
    
  </div>
</transition-stub>
`;
